### PR TITLE
Deploy to Heroku using Docker and Distillery

### PR DESCRIPTION
**Pull request by bokay**
We would like to have a working example of an application deployed on [Heroku](https://www.heroku.com) using [Docker](https://www.docker.com) and [Distillery](https://github.com/bitwalker/distillery). The application will require a Phoenix application with [Webpack](https://webpack.js.org) integration. 

Beyond using Docker for deployment, we would like to use the same Heroku/Docker setup locally. [+]

There are possibly multiple PR's here:

1. Phoenix application with Webpack
2. Configuration to generate a Release
3. Docker Deployment + Local Deployment

## References

- https://cultivatehq.com/posts/elixir-distillery-umbrella-docker/
- https://devcenter.heroku.com/articles/local-development-with-docker-compose [+]

### CODE REVIEW

**Comment by frahugo** [comment]
@bokay this is awesome!  Great job.  I tried it locally.  A few things I found:

- [ ] I followed the instructions in the readme.  When I run the app in Docker, it do not have CSS files available to the browser.  Maybe we need to cache an additional assets-related folder?
- [ ] This is more of a question for all of us: should we also support a local dev environment as well?  If so, the database config would then be different.  What would happen to the _build folder?  could we have compilation problems between docker architecture and MacOSX?
- [ ] If I do a `docker-compose up`, I cannot use the IEx console.  But I can if I run `docker-compose run application iex -S mix`.  Maybe the `up` command should simply start the Phoenix server instead of the console?
- [ ] `apps/magasin_web/assets/package-lock.json` is missing from the git repo

**Comment by frahugo** [comment]
- [ ] An error that I get when I run `docker run -i --name mag1 -t --rm -p 4000:4000 -e DATABASE_URL="postgres://user:password@hostname:5432/db" magasin/app /app/bin/magasin foreground`:

```
magasin git:(docker) ✗ docker run -i --name mag1 -t --rm -p 4000:4000 -e DATABASE_URL="postgres://user:password@hostname:5432/db" magasin/app /app/bin/magasin foreground
docker: Error response from daemon: driver failed programming external connectivity on endpoint mag1 (ff3f496f9f3225257f6398b0d7c895a2a8e78af51f2c68d1e4f9312dfd538e17): Bind for 0.0.0.0:4000 failed: port is already allocated.
```

**Comment by frahugo** [comment]
Got this warning at some point:

```
warning: Using {:system, "DATABASE_URL"} for your :url configuration is deprecated.

Instead define an init/2 callback in your repository that sets
the URL accordingly from your system environment:

    def init(_type, config) do
      {:ok, Keyword.put(config, :url, System.get_env("DATABASE_URL"))}
    end

  lib/ecto/repo/supervisor.ex:55: Ecto.Repo.Supervisor.compile_config/2
  lib/magasin/repo.ex:2: (module)
  (elixir) src/elixir_compiler.erl:85: :elixir_compiler.dispatch/6
  (elixir) src/elixir_module.erl:223: :elixir_module.eval_form/6
  (elixir) src/elixir_module.erl:77: :elixir_module.compile/5
```

**Comment by bokay** [comment]
@frahugo https://github.com/civilcode/magasin/pull/4#issuecomment-393166111 
You might have something listening to the port 4000 already. I suppose that your `docker-compose` is still up.

**Comment by frahugo** [comment]
- [ ] If I run `docker container ls -a`, I see a lot of `magasin_application_run_X` containers, X being a number that increments.  Is there a way to avoid that?

**Comment by frahugo** [comment]
@bokay I ran `docker-compose down`.  Now I get this error:

```
ocker run -i --name mag1 -t --rm -p 4000:4000 -e DATABASE_URL="postgres://user:password@hostname:5432/db" magasin/app /app/bin/magasin foreground
docker: Error response from daemon: Conflict. The container name "/mag1" is already in use by container "2b3377ae030382e7b1072c06d104b933322fa46005908d0cc664af3080f69298". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
```

**Comment by frahugo** [comment]
- [ ] When I run locally the prod environment, an error shows up in the output:

```
Running MagasinWeb.Endpoint with Cowboy using http://0.0.0.0:4000

13:57:05.077 [error] Could not find static manifest at "/app/lib/magasin_web-0.1.0/priv/static/manifest.json". Run "mix phx.digest" after building your static files or remove the configuration from "config/prod.exs".
```

**Comment by frahugo** [comment]
- [ ] When I run `docker build --no-cache -t magasin/app -f Dockerfile .`, I get this error:

```
Step 17/27 : RUN apk upgrade --no-cache &&     apk add --no-cache bash openssl
 ---> Running in e201a1fb2f87
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/x86_64/APKINDEX.tar.gz
(1/3) Upgrading busybox (1.26.2-r9 -> 1.26.2-r11)
Executing busybox-1.26.2-r11.post-upgrade
(2/3) Installing libressl2.5-libtls (2.5.5-r0)
ERROR: libressl2.5-libtls-2.5.5-r0: BAD signature
(3/3) Installing ssl_client (1.26.2-r11)
Executing busybox-1.26.2-r11.trigger
1 error; 4 MiB in 12 packages
The command '/bin/sh -c apk upgrade --no-cache &&     apk add --no-cache bash openssl' returned a non-zero code: 1
```

**Comment by bokay** [comment]
@frahugo You can give it another shot!

**Comment by bokay** [comment]
@nicholasjhenry @frahugo This PR is ready for review and merge!

File: guides/setup/docker.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ config :magasin_web, MagasinWeb.Endpoint,
+  http: [port: {:system, "PORT"}],
+  url: [host: "localhost", port: {:system, "PORT"}], # This is critical for ensuring web-sockets properly authorize.
+  cache_static_manifest: "priv/static/cache_manifest.json",
+  server: true,
+  root: ".",
+  version: Application.spec(:magasin_web, :vsn)
+
+```
+- [ ] Setup `prod.exs` to not import a `prod.secrets.exs`
```
Comment (nicholasjhenry) [inline guides/setup/docker.md]: Add references section as well, e.g. https://cultivatehq.com/posts/elixir-distillery-umbrella-docker/

Comment (bokay) [reply guides/setup/docker.md]: Reference added in https://github.com/civilcode/magasin/pull/4/commits/924f623d2ebc1ed44adab4a11fe962eb3a4b001f

**Pull request merged**